server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregator.java

```diff
@@ -150,118 +111,37 @@ public InternalAggregation buildEmptyAggregation() {
 
     @Override
     protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
-        if (leaf != null) {
-            leaf.docIdSet = builder.build();
-            contexts.add(leaf);
-        }
-        leaf = new LeafContext(ctx, sub);
-        builder = new RoaringDocIdSet.Builder(ctx.reader().maxDoc());
-        final CompositeValuesSource.Collector inner = array.getLeafCollector(ctx, getFirstPassCollector());
-        return new LeafBucketCollector() {
-            @Override
-            public void collect(int doc, long zeroBucket) throws IOException {
-                assert zeroBucket == 0L;
-                inner.collect(doc);
-            }
-        };
-    }
-
-    @Override
-    protected void doPostCollection() throws IOException {
-        if (leaf != null) {
-            leaf.docIdSet = builder.build();
-            contexts.add(leaf);
+        if (sortedBucketProducer != null && sortedBucketProducer.isApplicable(context.query())) {
+            // we can bypass search, the producer will visit documents sorted by the leading source
+            // of the composite definition and terminates when the leading source value is guaranteed to be
+            // greater than the lowest composite bucket in the queue.
+            sortedBucketProducer.processLeaf(context.query(), ctx, sub);
+            throw new CollectionTerminatedException();
```

Comment:
I wonder if it would be clearer for the SortedBucketProducer to throw the exception rather than throwing it here?

Reply:
I pushed a commit that adds a comment regarding why we throw an exception here.